sdl2promela/translator.py

```diff
@@ -0,0 +1,230 @@
+import typing
+from typing import List
+from multipledispatch import dispatch
+
+from opengeode import ogAST
+from opengeode.AdaGenerator import SEPARATOR
+
+from .sdl import model as sdlmodel
+from .promela import model as promelamodel
+from .promela.modelbuilder import (
+    AssignmentBuilder,
+    CallBuilder,
+    InlineBuilder,
+    VariableDeclarationBuilder,
+)
+from .promela.modelbuilder import VariableReferenceBuilder
+from .promela.modelbuilder import BlockBuilder
+from .promela.modelbuilder import ModelBuilder
+from .promela.modelbuilder import DoBuilder
+from .promela.modelbuilder import BinaryExpressionBuilder
+from .promela.modelbuilder import AlternativeBuilder
+from .promela.modelbuilder import SwitchBuilder
+
+
+from sdl2promela import promela
+
+__TRANSITION_TYPE_NAME = "int"
+__TRANSITION_ARGUMENT = "id"
+__TRANSITION_ID = "transition_id"
+__INVALID_ID = "-1"
+__STATE_VARIABLE = "state"
+__INIT = "init"
+
+
+def __get_transition_function_name(sdl_model: sdlmodel.Model) -> str:
+    return sdl_model.process_name + SEPARATOR + "transition"
+
+
+def __get_input_function_name(sdl_model: sdlmodel.Model, input: sdlmodel.Input) -> str:
+    return sdl_model.process_name + SEPARATOR + "PI" + SEPARATOR + input.name
+
+
+def __get_init_function_name(sdl_model: sdlmodel.Model) -> str:
+    return sdl_model.process_name + SEPARATOR + __INIT
+
+
+def __get_state_variable_name(sdl_model: sdlmodel.Model) -> str:
+    return sdl_model.process_name + SEPARATOR + __STATE_VARIABLE
+
+
+def __get_state_name(sdl_model: sdlmodel.Model, state: sdlmodel.State) -> str:
+    return sdl_model.process_name + SEPARATOR + state.name
+
+
+def __get_remote_function_name(
+    sdl_model: sdlmodel.Model, output: sdlmodel.Output
+) -> str:
+    return sdl_model.process_name + SEPARATOR + "RI" + SEPARATOR + output.name
+
+
+def __generate_input_function(
+    sdl_model: sdlmodel.Model, input: sdlmodel.Input
+) -> promelamodel.Inline:
+    builder = InlineBuilder()
+    builder.withName(__get_input_function_name(sdl_model, input))
+    for parameter in input.parameters:
+        builder.withParameter(parameter.name)
+    blockBuilder = BlockBuilder(promelamodel.BlockType.BLOCK)
+
+    switch_builder = SwitchBuilder()
+
+    state_variable_name = __get_state_variable_name(sdl_model)
+    transition_function_name = __get_transition_function_name(sdl_model)
+
+    for index, state in input.transitions.items():
+        switch_builder.withAlternative(
+            AlternativeBuilder()
+            .withCondition(
```

Comment:
although I do love Builder pattern, I wonder if in Python it wouldn't be easier to just rely on named arguments
```
Alternative(condition=....
```

but maybe then indents etc. would look worse here... just wondering

(trick for enforcing use of named arguments:
```
def fun(self, **kwargs):
  self.__fun(**kwargs)

def __fun(self, named1, named2, named3):
```

but that kills type checks :/ so maybe indeed Builder is a better way
)

Reply:
So do you propose a complete rewrite? I don't know what to do with this comment.